build: add a depends dependency provider

willcl-ark opened the pull request:
Fixes: #32428

This PR adds a [dependency provider](https://cmake.org/cmake/help/latest/guide/using-dependencies/index.html#dependency-providers) to depends builds.

Currently the depends toolchain appears to rely on certain assumptions [citation needed] to find packages correctly. It does largely do this correctly, however NixOS sets some environment variables which interfere with this and cause builds from depends to fail, due to attempting to link system packages.

We can reproduce this interferrence by running a NixOS container and loading system packages using a flake:

```bash
docker run --pull=always -it nixos/nix

# In the container:
git clone --depth=1 https://github.com/bitcoin/bitcoin && cd bitcoin

# Install system packages for a non-depends build from https://github.com/bitcoin-dev-tools/bix
nix develop github:bitcoin-dev-tools/bix --extra-experimental-features flakes --extra-experimental-features nix-command --no-write-lock-file

# Attempt depends build
make -C depends -j$(nproc)
cmake -B build --toolchain /bitcoin/depends/<host-platform-triplet>/toolchain.cmake

# Find possibly-interferring env vars:
env | grep -E 'CMAKE|NIX'
```

A dependency provider allows the overriding of  `cmake`'s `find_package()` therefore giving total control over where dependencies come from in a build.

This achieves two things:

1. Provides stronger guarantees about where dependencies come from during a (depends) build; not permitting environment flags to override these locations [citation needed].
2. Fixes issues like a non-standard `CMAKE_PREFIX_PATH` breaking builds (i.e. 32428)

This provider must be specified along with the toolchain, with an invocation of the type:

```bash
make -C depends
cmake -B build --toolchain depends/x86_64-pc-linux-gnu/toolchain.cmake -DCMAKE_PROJECT_TOP_LEVEL_INCLUDES=depends/x86_64-pc-linux-gnu/dependency_provider.cmake
cmake --build build
```
## To-do

- [x] Figure out why the provider does not like to find boost ([link](https://github.com/bitcoin/bitcoin/pull/32595/commits/f89253f9d3382d7cad084a11aff243e1e7d6bf5e)) fixed by rebasing on #32665 and #32667
- [x] Fix the failing CI job. This appears to somehow use an old `cmake`, even after we install `cmake` 4.0.2. Fixed by using bookworm, gcc-12 and `-Wno-error=array-bounds` in the previous releases job.

theuni committed:
depends: Bump boost to 1.88.0 and use new CMake buildsystem

This has a few advantages over the old method of simply copying headers:
- Installs proper cmake files which can be picked up by our buildsystem
- Only installs necessary headers, not all of boost

Co-authored-by: Hennadii Stepanov <32963518+hebasto@users.noreply.github.com>

willcl-ark committed:
build: add a depends dependency provider

willcl-ark commented:
Opened in draft for feedback on approach. 

There is also a bypass for the boost package (in _cmake/module/AddBoostIfNeeded.cmake_) which I'd like to get working with the dependency provider ideally.

Note that this also bumps the minimum cmake version to 3.24.

hebasto commented:
cc @purpleKarrot

fanquake commented:
> Provides stronger guarantees about where dependencies come from during a (depends) build; i.e. only from depends.

Can you elaborate on this? Currently, CMake configured with a depends toolchain, should never look outside depends (doing so would be a bug). What are the stronger guarantees?

willcl-ark commented:
> > Provides stronger guarantees about where dependencies come from during a (depends) build; i.e. only from depends.
> 
> Can you elaborate on this? Currently, CMake configured with a depends toolchain, should never look outside depends (doing so would be a bug). What are the stronger guarantees?

Right, perhaps I didn't use the best wording here.

When using the toolchain I _have not_ found us pulling in extraneous deps from outside of depends, but _have_ encountered us missing dependencies from inside of depends, e.g. in my original issue.

By using a dependency provider and intercepting calls to `find_package()`, as I understand it, we are getting even more control over which packages can be found than when using the toolchain.

This is what I termed "stronger guarantees", perhaps misleadingly!

willcl-ark commented:
The windows job fails as currently we activate the dependency provider automatically based on whether a(ny) toolchain is in use.

This works fine for depends, but doesn't make sense on windows where the vcpkg toolchain is used, without depends.

The idea for doing it this way was to try and avoid users having to pass _two_ flags in order to build from depends -- a toolchain and a dependency-provider -- but perhaps this is necessary after all...

I do recall reading somewhere in cmake docs that choice of dependency provider should always be left to the user too, so I suppose doing it this way would fit with that.

willcl-ark commented:
Pushed an update which now only activates the dependency provider when using the depends toolchain.

This will hopefully appease the windows CI job, as well as making more sense in general and being more expected behaviour.

purpleKarrot commented:
Thank you for working on a dependency provider!

I think core should provide *two* dependency providers, which will give users *three* different ways to build:

1. Not using a dependency provider: CMake will produce an error when dependencies are not found on the system. This will be the preferred way for distro packagers.
2. A fallback provider: This will use system packages when they are found and fall back to compiling vendored depends. Using this provider, `find_package` will always succeed, but it will not compile any "unnecessary" code. This approach may be preferred by developers or for quick experiments.
3. A depends provider: This will ignore system packages and always use vendored depends. This approach will be used for reproducible builds.

At the moment, the dependency provider points to prebuilt depends. Alternatively, the dependency provider could take the responsibility of building depends at the moment they are requested. This would make the current, Makefile based depends system an implementation detail that can be ported piece by piece to CMake.

willcl-ark commented:
Thanks for the feedback @purpleKarrot.

I like the sound of your approach here myself and will be happy to adapt to it. I think i'll start off by targetting 3.:

> A depends provider: This will ignore system packages and always use vendored depends. This approach will be used for reproducible builds.

...without the optionl "auto-build depends" functionality. This is close to what I have here already (and solves my immediate personal itch regarding building using depends on NixOS).

I would appreciate your input though on one more aspect of this; namely the (dev) UX involved with the 3 options. Option 1 would not see any UX changes.

Option 2 would require dev to specify the "fallback provider" file with a `-DCMAKE_PROJECT_TOP_LEVEL_INCLUDES` flag.

Option 3 would require the dev to specify **both** a `--toolchain` and a `-DCMAKE_PROJECT_TOP_LEVEL_INCLUDES`.

Option 3 in particular feels a little clunky to me (as a cmake novice), as I'd expect the toolchain to "contain everything needed to build". But I am guessing this can be abstracted away with a `CMakePreset` handling passing of the toolchain and dep. provider. This would become even more seamless in a version which auto-built depends for you too.

Does my understanding above sound roughly logical to you? If so I can make the changes later this week.

fanquake commented:
> Yes. The idea is to provide an approach to build core with minimal effort. 

I think `make -C depends && cmake -B --toolchain` and `apt install your_dependencies && cmake -B` are both already pretty minimal effort. It's not really clear why a developer would `apt install x`, but not `apt install x y`, and then would need a behaviour to fallback to depends for `y`?

I'm also not currently aware of a platform that we support, where the dependencies needed to compile Core, aren't available via the system package manager; one used to be our pecific version of BDB `4.x`, but that dependency has now been removed. So this fallback usecase mostly seems like additional build complexity, which creates a larger matrix of "supported" builds to test / debug, for not much gain. We used to have a similar kind of option in our depends system (`ALLOW_HOST_PACKAGES`), but allowing the use of system packages, with depends, defeats the point of a self-contained dependency builder, so it was removed.

There's also some nuance around how this would function, are you imaging it all integrated into into CMake, so depends is no-longer a stand alone system? Otherwise, we need logic in CMake, to detect if the depends requirements (curl, compilers, patch, binutils etc) are actually installed / usable (i.e otherwise the fallback would fail because curl isn't installed, probably leaving the user wondering why it's trying to download stuff), as well as making sure that build flags / compiler selection are all passed through to depends (i.e if the user is trying to build with Clang & libc++, then any tools / flags need to be forwarded to the depends compilation).

willcl-ark committed:
build: declare dependency provider from depends

After depends have been built, print out the required flags needed to
use the dependency provider along with the existing --toolchain message.

Include configure output showing active toolchain and dependency
provider information.

fanquake commented:
https://cirrus-ci.com/task/5352790488252416?logs=ci#L1675:
```bash
[09:31:48.288] Extracting freetype...
[09:31:48.316] /ci_container_base/depends/sources/freetype-2.11.0.tar.xz: OK
[09:31:48.562] Preprocessing freetype...
[09:31:48.575] Configuring freetype...
[09:31:48.656] CMake Error at CMakeLists.txt:100 (cmake_minimum_required):
[09:31:48.656]   Compatibility with CMake < 3.5 has been removed from CMake.
[09:31:48.656] 
[09:31:48.656]   Update the VERSION argument <min> value.  Or, use the <min>...<max> syntax
[09:31:48.656]   to tell CMake that the project requires at least <min> but has been updated
[09:31:48.656]   to work with policies introduced by <max> or earlier.
[09:31:48.656] 
[09:31:48.656]   Or, add -DCMAKE_POLICY_VERSION_MINIMUM=3.5 to try configuring anyway.
[09:31:48.656] 
[09:31:48.656] 
[09:31:48.656] -- Configuring incomplete, errors occurred!
```

hebasto committed:
build: Find Boost in config mode

The `FindBoost` module has been removed by policy CMP0167.

willcl-ark committed:
ci: use bookworm in native_previous_releases

Ubuntu 22.04 supplies cmake 3.22 which is below our minimum version.

- Switch to bookworm for cmake 3.25.
- Set FORTIFY_SOURCE=2, as level 3 creates a wraning on this platform.

willcl-ark commented on ci/test/01_base_install.sh:
```diff
@@ -32,6 +32,18 @@ if [ -n "${APT_LLVM_V}" ]; then
   )
 fi
 
+# Fetch newer cmake on Ubuntu 22.04
```
Hmmm, on bookworm (+ gcc12) we seem to fail on a new error: https://github.com/bitcoin/bitcoin/pull/32595/checks?check_run_id=44008178986 which I haven't looked into much yet

fanquake replied:
>  fail on a new error:

This looks similar to what we are already working around in the native fuzz valgrind task, so it seems fine to use the same workaround here (`-Wno-error=array-bounds`).

willcl-ark commented on ci/test/00_setup_env_native_previous_releases.sh:
```diff
@@ -7,10 +7,10 @@
 # Use minimum supported python3.10 and gcc-11, see doc/dependencies.md
-export PACKAGES="gcc-11 g++-11 python3-zmq"
-export DEP_OPTS="CC=gcc-11 CXX=g++-11"
+export PACKAGES="gcc-12 g++-12 python3-zmq"
```
This throws an error ~ that `FORTIFY_SOURCE=3` is a warning on gcc-11, and we are treating warnings as errors. I could use FORTIFY_SOURCE=2 in this job as a workaround that (which is why I moved to gcc-12 initially here)?

willcl-ark replied:
I see the same warning (error) locally on bookworm with gcc-11 too:

<details>
<summary>Details</summary>

```
C++ compiler .......................... GNU 11.3.0, /usr/bin/g++-11
CMAKE_BUILD_TYPE ...................... Debug
Preprocessor defined macros ........... DEBUG DEBUG_LOCKORDER DEBUG_LOCKCONTENTION RPC_DOC_CHECK ABORT_ON_FAILED_ASSUME
C++ compiler flags .................... -funsigned-char -g2 -O2 -std=c++20 -fPIC -fno-extended-identifiers -fdebug-prefix-map=/ci_container_base/src=. -fmacro-prefix-map=/ci_container_base/src=. -fstack-reuse=none -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -Wstack-protector -fstack-protector-all -fcf-protection=full -fstack-clash-protection -Werror -Wall -Wextra -Wformat -Wformat-security -Wvla -Wredundant-decls -Wdate-time -Wduplicated-branches -Wduplicated-cond -Wlogical-op -Woverloaded-virtual -Wsuggest-override -Wimplicit-fallthrough -Wunreachable-code -Wundef -Wno-unused-parameter -DBOOST_MULTI_INDEX_ENABLE_SAFE_MODE
Linker flags .......................... -funsigned-char -g2 -O2 -fstack-reuse=none -fstack-protector-all -fcf-protection=full -fstack-clash-protection -Wl,-z,relro -Wl,-z,now -Wl,-z,separate-code -Wl,--exclude-libs,ALL -fPIE -pie

NOTE: The summary above may not exactly match the final applied build flags
      if any additional CMAKE_* or environment variables have been modified.
      To see the exact flags applied, build with the --verbose option.

Treat compiler warnings as errors ..... ON
Use ccache for compiling .............. ON


-- Configuring done
-- Generating done
-- Build files have been written to: /ci_container_base/ci/scratch/build-x86_64-pc-linux-gnu
+ bash -c 'cmake --build . -j15 --target all install'
[  0%] Building CXX object src/CMakeFiles/crc32c.dir/crc32c/src/crc32c.cc.o
[  0%] Building CXX object src/univalue/CMakeFiles/univalue.dir/lib/univalue.cpp.o
[  0%] Building C object src/secp256k1/src/CMakeFiles/secp256k1_precomputed.dir/precomputed_ecmult.c.o
[  0%] Building CXX object src/CMakeFiles/bitcoin_cli.dir/compat/stdin.cpp.o
[  0%] Building CXX object src/crypto/CMakeFiles/bitcoin_crypto.dir/aes.cpp.o
[  0%] Building CXX object src/univalue/CMakeFiles/univalue.dir/lib/univalue_get.cpp.o
[  0%] Building C object src/secp256k1/src/CMakeFiles/exhaustive_tests.dir/tests_exhaustive.c.o
[  2%] Building CXX object src/test/util/CMakeFiles/test_util.dir/blockfilter.cpp.o
[  2%] Building CXX object src/CMakeFiles/minisketch.dir/minisketch/src/minisketch.cpp.o
[  2%] Generating bitcoin-build-info.h
[  2%] Building CXX object src/CMakeFiles/bitcoin_consensus.dir/arith_uint256.cpp.o
[  2%] Building CXX object src/univalue/CMakeFiles/univalue.dir/lib/univalue_read.cpp.o
[  2%] Building CXX object src/zmq/CMakeFiles/bitcoin_zmq.dir/zmqabstractnotifier.cpp.o
[  2%] Building CXX object src/CMakeFiles/leveldb.dir/leveldb/db/builder.cc.o
[  2%] Generating locale/bitcoin_am.qm
[  2%] Generating locale/bitcoin_ar.qm
[  2%] Generating locale/bitcoin_ast_ES.qm
In file included from /usr/include/x86_64-linux-gnu/c++/11/bits/os_defines.h:39,
                 from /usr/include/x86_64-linux-gnu/c++/11/bits/c++config.h:586,
                 from /usr/include/c++/11/cstdio:41,
                 from /ci_container_base/src/compat/stdin.cpp:7:
/usr/include/features.h:423:5: error: #warning _FORTIFY_SOURCE > 2 is treated like 2 on this platform [-Werror=cpp]
  423 | #   warning _FORTIFY_SOURCE > 2 is treated like 2 on this platform
      |     ^~~~~~~
cc1plus: all warnings being treated as errors
gmake[2]: *** [src/CMakeFiles/bitcoin_cli.dir/build.make:76: src/CMakeFiles/bitcoin_cli.dir/compat/stdin.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:880: src/CMakeFiles/bitcoin_cli.dir/all] Error 2
gmake[1]: *** Waiting for unfinished jobs....
```

</details>

I added [-Wno-error=array-bounds](https://github.com/bitcoin/bitcoin/pull/32595#discussion_r2144445902) to `CMAKE_CXX_FLAGS` as in fuzz job, but this just still complained about "the fortify source level on this platform".

Going to reduce the FORTIFY_SOURCE level for now in absence of another fix.

fanquake replied:
> I added https://github.com/bitcoin/bitcoin/pull/32595#discussion_r2144445902 to CMAKE_CXX_FLAGS

That workaround is only for GCC 12.

willcl-ark replied:
taken -Wno-error=array-bounds and gcc-12 plus bookworm

willcl-ark commented on CMakeLists.txt:
```diff
@@ -7,7 +7,7 @@
 # Centos Stream 9, https://www.centos.org/cl-vs-cs/#end-of-life, EOL in May 2027:
 #  - CMake 3.26.5, https://mirror.stream.centos.org/9-stream/AppStream/x86_64/os/Packages/
-cmake_minimum_required(VERSION 3.22)
+cmake_minimum_required(VERSION 3.24)
```
Going to leave this at 3.24 for now

josibake reviewed (commented):
Thanks for working on this, @willcl-ark ! I got nerd sniped on this and ended up learning a bunch about CMake, toolchains, and dependency providers. 

As written now, I don't think this PR actually solves #32428, so per https://github.com/bitcoin/bitcoin/issues/32428#issuecomment-2991258328, I do think for nix / nixos specifically we'd be better off going the simple route for now, and then continuing to work on this PR as a more comprehensive solution.

willcl-ark commented on depends/dependency_provider.cmake:
```diff
@@ -0,0 +1,28 @@
+    set(${package_name}_INCLUDE_DIR "${DEPENDS_DIR}/include" CACHE PATH "" FORCE)
+    set(${package_name}_INCLUDEDIR "${DEPENDS_DIR}/include" CACHE PATH "" FORCE)
+    set(${package_name}_LIBRARY_DIR "${DEPENDS_DIR}/lib" CACHE PATH "" FORCE)
+    find_package(${package_name} ${ARGN} BYPASS_PROVIDER)
```
Correct. Currently this provider is only configuring the correct directories dynamically on `find_package()` calls, then "bypassing" fully over-riding `find_package()`. So essentially it's more of a "configuration & passthrough provider".

I'm going to close this for now (could mark as up for grabs?) as a) this doesn't achieve any increased hermetic properties as it stands, and b) I agree with the comments in #32428 that we should implement a simple fix for Nix users first, with the aim of introducing a more full-featured provider in the future.

(This was actually my preferred approach, hence me including just such a [patch](https://github.com/bitcoin/bitcoin/commit/437a4da93f437f9588ff32d19ac59fea3802c676) in the original issue comment in https://github.com/bitcoin/bitcoin/issues/32428#issue-3043229696
